stats/stats/src/charts/types/timespans/week.rs

```diff
@@ -0,0 +1,254 @@
+            b.field(&"invalid");
+        }
+        b.finish()
+    }
+}
+
+// `NaiveWeek` doesn't implement common traits, we have to do it by hand
+
+impl PartialEq for Week {
+    fn eq(&self, other: &Self) -> bool {
+        // if both `None` then it's `NaiveDate::MIN`'s week,
+        // thus equal
+        self.checked_first_day().eq(&other.checked_first_day())
+    }
+}
+
+impl Eq for Week {}
+
+impl PartialOrd for Week {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Week {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // `None` means `NaiveDate::MIN`'s week
+        // which fits well with `Option`'s "`None` is less than `Some`"
+        // policy
+        self.checked_first_day().cmp(&other.checked_first_day())
+    }
+}
+
+impl Clone for Week {
+    fn clone(&self) -> Self {
+        // Saturation happens only in one week
+        Self::new(self.saturating_first_day())
+    }
+}
+
+impl Timespan for Week {
+    fn from_date(date: NaiveDate) -> Self {
+        Week::new(date)
+    }
+
+    fn into_date(self) -> NaiveDate {
+        self.saturating_first_day()
+    }
+
+    fn enum_variant() -> ResolutionKind {
+        ResolutionKind::Week
+    }
+
+    fn saturating_start_timestamp(&self) -> chrono::DateTime<chrono::Utc> {
+        self.saturating_first_day().saturating_start_timestamp()
+    }
+
+    fn saturating_add(&self, duration: TimespanDuration<Self>) -> Self
+    where
+        Self: Sized,
+    {
+        let result_week_date = self
+            .saturating_first_day()
+            .checked_add_days(Days::new(duration.repeats() * 7))
+            .unwrap_or(NaiveDate::MAX);
+        Self::from_date(result_week_date)
+    }
+
+    fn saturating_sub(&self, duration: TimespanDuration<Self>) -> Self
+    where
+        Self: Sized,
+    {
+        let result_week_date = self
+            .saturating_first_day()
+            .checked_sub_days(Days::new(duration.repeats() * 7))
+            .unwrap_or(NaiveDate::MIN);
+        Self::from_date(result_week_date)
+    }
+}
+
+impl ConsistsOf<NaiveDate> for Week {
+    fn from_smaller(date: NaiveDate) -> Self {
+        Week::from_date(date)
+    }
+
+    fn into_smaller(self) -> NaiveDate {
+        Week::into_date(self)
+    }
+}
+
+impl Week {
+    pub fn new(day: NaiveDate) -> Week {
+        Self(day.week(WEEK_START))
+    }
+
+    /// `None` - out of range
+    pub fn checked_days(&self) -> Option<RangeInclusive<NaiveDate>> {
+        Some(self.checked_first_day()?..=self.checked_last_day()?)
+    }
+
+    /// `None` - out of range (week of `NaiveDate::MIN`)
+    pub fn checked_first_day(&self) -> Option<NaiveDate> {
+        // Current interface of `NaiveWeek` does not provide any way
+        // of idiomatic overflow handling
+        // todo: change after https://github.com/chronotope/chrono/pull/1600
+        // is released
+        std::panic::catch_unwind(|| self.0.first_day()).ok()
```

Comment:
i suggest remove panic logs, at least in production.

If there is no way to remove panic logs, lets write our own NaiveWeek

Reply:
fixed it with patch, problem with maintaining it should not appear, as these changes are expected to be added to the next release (if not, it will be discovered soon and I'll address this in some other way)